messages/src/mgmt/mod.rs

```diff
@@ -87,145 +86,317 @@ pub enum ManagementInterface {
+    /// Return the offset into the page of the start of the read.
+    pub fn offset(&self) -> u8 {
+        self.offset
     }
 
-    /// Return the bank of the upper page being accessed.
-    ///
-    /// See [`UpperPage::bank`] for details.
-    pub fn bank(&self) -> Option<u8> {
-        self.upper_page.bank()
+    /// Return the number of bytes read.
+    pub fn len(&self) -> u8 {
+        self.len
     }
+}
+
+/// A sized read access to a transceiver memory page.
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize, SerializedSize)]
+pub struct MemoryWrite {
+    page: Page,
+    offset: u8,
+    len: u8,
+}
 
-    /// Return the offset into the transceiver memory map for the operation.
+impl MemoryWrite {
+    /// Create a new validated write to the specific page.
+    pub fn new<P>(page: P, offset: u8, len: u8) -> Result<Self, Error>
+    where
+        P: MemoryPage + Into<Page>,
+    {
+        verify_write(&page, offset, len)?;
+        Ok(Self {
+            page: page.into(),
+            offset,
+            len,
+        })
+    }
+
+    /// Return the page being read.
+    pub fn page(&self) -> &Page {
+        &self.page
+    }
+
+    /// Return the offset into the page of the start of the write.
     pub fn offset(&self) -> u8 {
         self.offset
     }
 
-    /// Return the length of transceiver memory for the operation.
+    /// Return the number of bytes written.
     pub fn len(&self) -> u8 {
         self.len
     }
 }
 
+/// A trait describing the limits of a memory page.
+pub trait MemoryPage: Copy {
+    /// The maximum allowed offset into the memory page.
+    fn max_offset(&self) -> u8;
+
+    /// The minimum allowed offset into the memory page.
+    ///
+    /// Note that these values are always referenced to the full 256-byte
+    /// transceiver memory map. So this should be zero for a lower page, and 128
+    /// for an upper page.
+    fn min_offset(&self) -> u8 {
+        0
+    }
+
+    /// The upper limit of the page.
+    ///
+    /// Specifically, this is the offset of the first _invalid_ byte of the
+    /// page, or equivalently the maximum offset plus 1.
+    fn upper_limit(&self) -> u16 {
+        self.max_offset() as u16 + 1
+    }
+
+    /// The minimum size of a write operation, in bytes.
+    const MIN_WRITE_SIZE: u8 = 1;
+
+    /// The maximum size of a write operation, in bytes.
+    const MAX_WRITE_SIZE: u8;
+
+    /// The minimum size of a read operation, in bytes.
+    const MIN_READ_SIZE: u8 = 1;
+
+    /// The maximum size of a read operation, in bytes.
+    const MAX_READ_SIZE: u8;
+}
+
+// Helper to check that the limits provided define a valid read.
+fn verify_read<P>(page: &P, offset: u8, len: u8) -> Result<(), Error>
+where
+    P: MemoryPage,
+{
+    if offset >= page.min_offset()
+        && offset <= page.max_offset()
+        && (offset as u16 + len as u16) <= page.upper_limit()
+        && len >= P::MIN_READ_SIZE
+        && len <= P::MAX_READ_SIZE
```

Comment:
I stared at this a lot, and can't find any off-by-one errors!

Reply:
There's a first time for everything :)